### DISCUSSION TITLE
booting from ISO w/ embedded script not working

### DISCUSSION THREAD

**Discussion by tfordon** [ipxe/ipxe#497] [Q&A] [answered]
I'm having trouble creating an ISO with an embedded script that automatically runs.  I'm sure there is something I'm just not understanding.

I added a file called myscript.ipxe into the src folder with the following contents:
```
#!ipxe

echo HELLO
```

I then compiled it by running:
```
make bin/ipxe.iso EMBED=myscript.ipxe
```

I call QEMU using the ISO:
```
qemu-system-x86_64 -m 1024 bin/ipxe.iso
```

I get the following:
![image](https://user-images.githubusercontent.com/1718496/137546409-1214196c-e478-48ef-9cf0-a5f1d6a3a16b.png)

I thought myscript.ipxe would be executed and I would see "HELLO" somewhere on the screen.

I looked at usr/autoboot.c and saw it boots the first image if it is available, so I restarted QEMU and hit ctrl-b at the prompt.  Then I typed `imgstat` and saw the following:
![image](https://user-images.githubusercontent.com/1718496/137546715-27c2ddb7-cef3-40d7-aff8-4fc5450bfcd0.png)

I thought myscript.ipxe would be in the image list, but it looks empty.  Should I see something?

I ensured there is no BOM at the beginning of myscript.ipxe, based on an old message in the previous forum (https://forum.ipxe.org/showthread.php?tid=8138)

Thank you for any help,
Tom

**Comment by mcb30** [comment, marked as answer]
Your screenshot shows that QEMU is `Booting from ROM...`, i.e. using the built-in iPXE ROM that comes as part of QEMU.  This built-in iPXE ROM is unrelated to the iPXE image that you have tried to build yourself, which is why you are not seeing any evidence of your embedded script.

The problem is that you have built a CD-ROM image (`.iso`) but then told QEMU that it is a hard disk image.  You will need to either tell QEMU that the image is for a CD-ROM using the `-cdrom` option:
```
make bin/ipxe.iso EMBED=myscript.ipxe
qemu-system-x86_64 -m 1024 -cdrom bin/ipxe.iso
```
or, alternatively, build a hard disk image (`.usb`) instead:
```
make bin/ipxe.usb EMBED=myscript.ipxe
qemu-system-x86_64 -m 1024 bin/ipxe.usb
```

**Reply by tfordon** [reply]
Thanks for the very helpful answer.  I feel pretty silly.
